RavenDB-22110 CollectionQuery count

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22110

### Additional description

PR includes fixes to count, distinct, and paging for very specific cases.
In addition, it contains cleanup for the previous collection query flow to make it more straightforward.
 

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-22110 CollectionQuery
    - removed multiple enumerators in collection query
    - distinct query - correct query response metadata (e.g. TotalResults)
    - startsWith paging fix

ayende commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -250,8 +253,46 @@ public bool MoveNext()
+                if (_totalResultsCalculated && _query.Metadata.IsDistinct == false)
+                    return;
+                
+                // We need to disable calculating totalResults in the enumerator, so let's set _totalResultsCalculated to true.
```
Why are we counting manually? If we don't have distinct, can't we get that from the number of items without materializing everything?

maciejaszyk replied:
If we don't have distinct  clause and we know from storage how many items we have we are escaping materializing (line #265), however if we don't know how many items our enumerable has (this is for example `In`/`startsWith`) we've to go through all data. (We handled it before very similarly: https://github.com/ravendb/ravendb/blob/f632fa9ff761aa5d206eba57abe66ce48ffffc7a/src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs#L353-L359)

However, it still seems to be not correct for cases when we've filter. For such case we've to materialize, check via script and then count, right?

ayende replied:
With `filter`, that require to actually execute it all, I think

maciejaszyk committed:
RavenDB-22110
- Support filter queries for paging
- Support filter queries for count
- Reverted paging part for IN queries  (internal)
- Support for unbounded queries
- Memory leak fix for count / initialize (also for compressed database)
- Treat `in` as bounded query

arekpalinski commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -223,15 +220,30 @@ private class Enumerator : IEnumerator<Document>
                 if (_initialized == false)
-                    _start = Initialize();
+                {
+                    Initialize(out var processedAllDocuments);
+                    
+                    // When our query skipped all documents from underlying stream
+                    if (processedAllDocuments)
+                        return false;
```
I wonder whether just returning `false` from `Initialize()` wouldn't be better then:

```
if (Initialize() == false)
{
    // When our query skipped all documents from underlying stream
    return false;
}
```

maciejaszyk replied:
We want to avoid method calls here since it's hot-path.

maciejaszyk replied:
Since it's only one call per enumerator instance, I think it's better to return the value explicitly with the name. Returning it as a return parameter from `Initialize` can be confusing.

arekpalinski replied:
ok

arekpalinski commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -250,8 +262,60 @@ public bool MoveNext()
+                if (_totalResultsCalculated && _query.Metadata.IsDistinct == false && _filterScriptRun == null)
+                    return;
+                
+                // For count(): We need to disable calculating totalResults in the enumerator, so let's set _totalResultsCalculated to true.
```
Can you explain why we need to disable it then?

maciejaszyk replied:
For unbounded queries, we have to count it ourselves, but if we don't go through all the documents, it returns -1. So count is the case where we go and count everything. In order not to mix the logic between this method and enumerator, the count for `count()` is done manually in this method.

arekpalinski commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -250,8 +262,60 @@ public bool MoveNext()
+                        }
+                    }
+                    
+                    _context.Transaction.ForgetAbout(doc);
```
When I noticed that I thought about the usage of `TransactionForgetAboutEnumerator` but then I found we use it here:

https://github.com/ravendb/ravendb/blob/8fa577b15db528cdd1362e1feae3bb65e625e56d/src/Raven.Server/Documents/Queries/Dynamic/CollectionQueryRunner.cs#L164

But here it's a different case because we're just counting, right? Can we wrap the underlying enumerator by `TransactionForgetAboutEnumerator`?

maciejaszyk replied:
We can't wrap it because it's not an enumerator in the `IEnumerator' sense. We handle the state ourselves.

arekpalinski commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -332,36 +383,18 @@ public bool MoveNext()
                 IEnumerable<Document> documents;
-
+                var skip = amountToSkip ?? 0;
+                const long take = long.MaxValue; // owner of enumerable decide when to stop, there is no reason to make paging at storage level.
```
Maybe `take` -> `takeAll`?

maciejaszyk replied:
Done

arekpalinski commented on src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs:
```diff
@@ -383,100 +416,119 @@ private IEnumerable<Document> GetDocuments()
                         {
-                            return d.Data.Count > 0 && _alreadySeenProjections.Add(d.DataHash) && _alreadySeenProjections.Count == _query.Start;
+                            document.Dispose();
+                            _context.Transaction.ForgetAbout(document);
```
Similar concern like above regarding direct usage of `ForgetAbout()`

maciejaszyk committed:
RavenDB-22110 PR Comments

arekpalinski commented:
@maciejaszyk according to https://github.com/ravendb/ravendb/actions/runs/8519303544/job/23333064496?pr=18284 file `RavenDB_22110.cs` contains tabs (instead of spaces)
